### PR TITLE
add a describe method on DataFrame like Polars

### LINKED ISSUE #4974: Is there a describe method on DataFrame like Polars?
I was trying to generate a summary on the data frame but could not find any relevant methods like `describe` or `summary` which can provide [this kind](https://docs.rs/polars/latest/polars/frame/struct.DataFrame.html#example-45) of table on the data which Polars data frame has an implementation.

**Describe the solution you'd like**
A described the method on the data frame `&df.describe().await?` should return a new summary dataframe with mean, median, and max values for each column. In more advance case metric name can be provided as vec so that it only compute/statistically estimate tose values for example `&df.describe(vec!["min", "max", "25%"]).await?`

### PR DESCRIPTION
**Pull request by jiangzhx**
# Which issue does this PR close?

Closes #4974 .

### CODE REVIEW

**Comment by jiangzhx** [comment]
Try to complete this work, there are still many places that can be optimized.
image

**Comment by jiangzhx** [comment]
make describe method return right datatype for record batch。

image

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -302,6 +306,177 @@ impl DataFrame {
...
+            self.clone()
+                .aggregate(
+                    vec![],
+                    fields_iter
+                        .clone()
+                        .map(|f| {
+                            Expr::Alias(
+                                Box::new(datafusion_expr::count(col(f.name()))),
+                                f.name().to_string(),
+                            )
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:378]: I think you can write this more concisely like this (untested)

```suggestion
                            datafusion_expr::count(col(f.name()).alias(f.name())
```

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -302,6 +306,177 @@ impl DataFrame {
...
+    /// let ctx = SessionContext::new();
+    /// let df = ctx.read_csv("tests/tpch-csv/customer.csv", CsvReadOptions::new()).await?;    
+    /// df.collect_describe().await.unwrap();
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn collect_describe(self) -> Result<Vec<RecordBatch>> {
+        //the functions now supported
+        let supported_describe_functions = vec!["count", "null_count", "max", "min"]; //"count",  "max", "min",
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:349]: Is the `// count` comment on purpose?

File: datafusion/core/src/dataframe.rs [hunk 3]
```diff
@@ -302,6 +306,177 @@ impl DataFrame {
...
+        let fields_iter = self.schema().fields().iter();
+
+        //define describe column
+        let mut describe_schemas = fields_iter
+            .clone()
+            .map(|field| {
+                if field.data_type().is_numeric() {
+                    Field::new(field.name(), DataType::Float64, true)
+                } else {
+                    Field::new(field.name(), DataType::Utf8, true)
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:360]: I would expect that the schema for `count` and `null_count` were always `Int64` and the schema for min/max were always `Utf8`

File: datafusion/core/src/dataframe.rs [hunk 4]
```diff
@@ -302,6 +306,177 @@ impl DataFrame {
...
+                )?
+                .collect()
+                .await?,
+            // max aggregation
+            self.clone()
+                .aggregate(
+                    vec![],
+                    fields_iter
+                        .clone()
+                        .filter(|f| matches!(f.data_type().is_numeric(), true))
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:408]: I wonder why restrict the min/max aggregation to numeric fields?

In order to get the min/max values in all columns to work, you could call `cast` to cast them to the same datatype

File: datafusion/core/src/dataframe.rs [hunk 5]
```diff
@@ -302,6 +306,177 @@ impl DataFrame {
...
+    /// # async fn main() -> Result<()> {
+    /// let ctx = SessionContext::new();
+    /// let df = ctx.read_csv("tests/tpch-csv/customer.csv", CsvReadOptions::new()).await?;    
+    /// df.describe().await?;
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn describe(self) -> Result<()> {
+        Ok(print_batches(
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:326]: I think a nicer interface would be 

```rust
pub async fn describe(self) -> Result<Self> {
```

And then we don't need a separate `collect_describe` instead we could reuse the existing `collect()`

One way might be to return a new `DataFrame` from the created RecordBatch -- perhaps following the model of

https://docs.rs/datafusion/18.0.0/src/datafusion/execution/context.rs.html#719-730

**Review by alamb — COMMENTED** [review]
Thank you @jiangzhx  -- this looks like a very nice addition to DataFusion ❤️ 

I left some comments -- let me know if anything isn't clear.

And thanks again!

File: datafusion/core/src/dataframe.rs (hunk 1 above)
Comment (jiangzhx) [reply datafusion/core/src/dataframe.rs:378]: thanks,  This API is more convenient to use.

File: datafusion/core/src/dataframe.rs (hunk 2 above)
Comment (jiangzhx) [reply datafusion/core/src/dataframe.rs:349]: forgot to delete,fixed.

File: datafusion/core/src/dataframe.rs (hunk 4 above)
Comment (jiangzhx) [reply datafusion/core/src/dataframe.rs:408]: boolean and binary not work with min/max

Comment (jiangzhx) [reply datafusion/core/src/dataframe.rs:408]: date_string_col, string_col ’s datatype also Binary. 
called `Result::unwrap()` on an `Err` value: Internal("Min/Max accumulator not implemented for type Binary")

File: datafusion/core/src/dataframe.rs (hunk 5 above)
Comment (jiangzhx) [reply datafusion/core/src/dataframe.rs]: thanks , it's great suggestion.
fixed

File: datafusion/core/src/dataframe.rs (hunk 3 above)
Comment (jiangzhx) [reply datafusion/core/src/dataframe.rs:337]: the describe method return schema like this.
image


the each column should have same datatype .
for example :
- `bool_col`  on  `count/null_count` return Int64 ;  error on  `min/max` , so make `bool_col` datatype `UTF8`;  
- `float_col` on  `count/null_count` return Int64 ;  on  `min/max`  return float, so make `float_col` datatype `Float64`

File: datafusion/core/src/dataframe.rs [hunk 6]
```diff
@@ -302,6 +305,155 @@ impl DataFrame {
...
+
+        //insert first column with function names
+        array_ref_vec.insert(
+            0,
+            Arc::new(StringArray::from_slice(
+                supported_describe_functions.clone(),
+            )),
+        );
+
+        let describe_record_batch =
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:439]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @jiangzhx 

cc @andygrove  (as I think this is a neat thing to expose in datafusion-python)

**Pull request merged**